src/patterns.md

```diff
@@ -756,6 +761,57 @@ Path patterns are irrefutable when they refer to structs or an enum variant when
 has only one variant or a constant whose type is irrefutable. They are refutable when they
 refer to refutable constants or enum variants for enums with multiple variants.
 
+## Or-patterns
+
+_Or-patterns_ are patterns that match on either of two sub-patterns (e.g. `A |
+B`). They can nest arbitrarily. Syntactically, or-patterns are allowed in any
+of the places where other patterns are allowed (represented by the _Pattern_
+production), with the exceptions of function and closure arguments (represented
+by the _PatternNoTopAlt_ production). Additionally, the macro `:pat` matchers
+only match or-patterns in the 2021+ editions.
+
+### Static semantics
+
+1. Given a pattern `p | q` at some depth for some arbitrary patterns `p` and `q`,
+   the pattern is considered ill-formed if:
+
+   + the type inferred for `p` does not unify with the type inferred for `q`, or
+   + the same set of bindings are not introduced in `p` and `q`, or
+   + the type of any two bindings with the same name in `p` and `q` do not unify
+     with respect to types or binding modes.
+
+   [type coercions]: https://doc.rust-lang.org/reference/type-coercions.html
```

Comment:
Links should always be relative, and preferably placed at the bottom of the page.

```suggestion
   [type coercions]: type-coercions.md
```